src/input.rs

```diff
@@ -0,0 +1,497 @@
+    }
+
+    /// Checks if mapped action button is released.
+    pub fn action_released<T>(&self, action: T) -> bool
+    where
+        Self: ActionMapper<T>,
+        T: Copy + Eq + std::hash::Hash,
+    {
+        match self.action_state(action) {
+            Some(state) => state == State::Released,
+            None => false,
+        }
+    }
+
+    /// Checks if mapped button is pressed or hold.
+    pub fn action_hold<T>(&self, action: T) -> bool
+    where
+        Self: ActionMapper<T>,
+        T: Copy + Eq + std::hash::Hash,
+    {
+        match self.action_state(action) {
+            Some(state) => state == State::Hold || state == State::Pressed,
+            None => false,
+        }
+    }
+
+    /// Just a mapper getter.
+    pub fn get_mapper<T: 'static + Send + Sync>(&self) -> &T {
+        self.mapper.downcast_ref::<T>().unwrap()
+    }
+
+    /// Initialize input manager - it is requiered to get information about window size.
+    pub fn init(&mut self, window: &winit::window::Window) {
+        self.window_size = Vector2::new(window.inner_size().width as f32, window.inner_size().height as f32);
+    }
+
+    /// This method should be called in application.rs after render, so States from events will be properly updated.
+    pub fn handle_update(&mut self) {
+        self.last_mouse_position = self.mouse_position;
+        self.mouse_scroll_delta = 0.0;
+
+        self.states.retain(|_btn, state| match state {
+            State::Pressed => {
+                *state = State::Hold;
+                true
+            }
+            State::Released => false,
+            _ => true,
+        });
+    }
+
+    /// This method should be called in application.rs in event_loop, so systems will get actual Input data.
+    pub fn handle_event(&mut self, event: &winit::event::Event<'_, ()>) {
+        match event {
+            Event::WindowEvent {
+                event: WindowEvent::KeyboardInput { input, .. },
+                ..
+            } => self.handle_keyboard_event(input),
+            Event::WindowEvent {
+                event: WindowEvent::MouseInput { state, button, .. },
+                ..
+            } => self.handle_mouse_click_event(*state, *button),
+            Event::WindowEvent {
+                event: WindowEvent::CursorMoved { position, .. },
+                ..
+            } => self.handle_cursor_moved_event(position),
+            Event::WindowEvent {
+                event: WindowEvent::MouseWheel { delta, .. },
+                ..
+            } => self.handle_mouse_wheel_event(&delta),
+            Event::WindowEvent {
+                event: WindowEvent::Resized(size),
+                ..
+            } => self.window_size = Vector2::new(size.width as f32, size.height as f32),
+            _ => {}
+        }
+    }
+
+    /// Handle cursor moved event from winit.
+    pub fn handle_cursor_moved_event(&mut self, position: &winit::dpi::PhysicalPosition<f64>) {
+        self.mouse_position = Vector2 {
```

Comment:
shouldn't it be here a sum of vectors?

Reply:
I think it work's fine. Let's try demo - hold left mouse button and move with mouse. Mouse position is printed to console